services/alertmanager/registry.go

```diff
@@ -17,11 +17,21 @@
 package alertmanager
 
 import (
+	"os"
 	"strings"
 	"sync"
 	"time"
 
+	"github.com/go-openapi/strfmt"
 	"github.com/percona/pmm/api/alertmanager/ammodels"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// Environment variable to overwrite resendInterval during testing
+	envResendInterval     = "PERCONA_TEST_ALERTMANAGER_RESEND_INTERVAL"
+	resolveTimeoutFactor  = 3
+	defaultResendInterval = 30 * time.Second
```

Comment:
I think `envResendInterval` and `defaultResendInterval`  variables should be declared in `alertmanager.go` as resend logic declared there.

Reply:
but they are used in `registry.go` so it didn't make sense for me to move them out to `alertmanager.go`, let me know if you still insist on doing this and I will make the change

Reply:
if i recall correctly in your draft PR you were using resend interval and env var in your `alertmanager.go` code, that is why it made sense to put them there, but here I have moved these to registry because Alexy wanted to make `resendInterval` a `registry` field  https://github.com/percona/pmm-managed/pull/405#discussion_r438031028

Reply:
My point is that `resendInterval` is not `registry` value, as the registry doesn't resend anything. The registry requires another value, `alert TTL` that depends on `resendInterval` value.  In my draft PR both variables are fields of corresponding services, but also there is a constant `defaultResendInterval`. So both serviced just check env variable to figure out was `defaultResendInterval` overwritten or not.  If it's not, then alertmanager uses `defaultResendInterval` as `resendInterval` and `registry` uses `defaultResendInterval` to calculate `alertTTL` (`resolveTimeoutFactor * defaultResendInterval`). Otherwise both use value of env variable for that.

For me, such approach looks more natural, as all variables declared in expected files. 

@AlekSi what do you think?  

p.s. here is the draft we discussing: https://github.com/percona/pmm-managed/pull/412/files

Reply:
While I agree with most of it, it sounds very concerning that we use _environment variable_ to pass information in code. Not function parameters, method parameters, global constants, etc – envrionment variables. Something that compiler and linters can't check.

I don't know what is the best way to solve this problem, but using "global" environenment variable is not it.

I wonder if changing the scope of "registry" from "it stores all generic alerts" to "it stores current STT alerts" and moving it to checks package, and then passing alerts from check's registry to alertmanager problem will solve the problem. Alert type definition (without Alertmanager and STT details, generic enough) can be in shared `services` package.

Reply:
> I wonder if changing the scope of "registry" from "it stores all generic alerts" to "it stores current STT alerts" and moving it to checks package, and then passing alerts from check's registry to alertmanager problem will solve the problem. Alert type definition (without Alertmanager and STT details, generic enough) can be in shared services package.

so should we try doing this as a part of this ticket or a new ticket?

Reply:
Let's quickly try it in this ticket

Reply:
> I wonder if changing the scope of "registry" from "it stores all generic alerts" to "it stores current STT alerts" and moving it to checks package, and then passing alerts from check's registry to alertmanager problem will solve the problem. Alert type definition (without Alertmanager and STT details, generic enough) can be in shared services package.

@aleksi I am not sure how this would help us eliminate the use of `resendInterval` as an env var (can you elaborate on this?), even if I move the registry to `checks` package, I will need a `resendInterval` value either as a global constant or env var to set the `EndsAt` field. Why not just keep the code where it is and pass `resendInterval` as a parameter to `NewRegistry`

Reply:
That should allow us to un-export `registry` type and make it an implementation detail. In turn, that should allow us to make `checks` and `alertmanager` services less coupled

Reply:
> That should allow us to un-export registry type

Currently we are exporting `Registry` `alertmanager` to make it visible in the `checks` package but if we move `Registry` to `checks` we will have to export it anyway to make it visible in `alertmanager` @AlekSi

Reply:
Let’s continue with the fastest way for now. It looks like we will have to refactor this part anyway. I have an idea how to do that, but it will take more time